core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java

```diff
@@ -137,8 +138,52 @@ public CalciteCatalogReader withSchemaPath(List<String> schemaPath) {
     return config;
   }
 
+  /**
+   * Gets the macro with the given name if it exists.
+   *
+   * @param names The name of the macro, list represents possibility of a
+   *              compound identifier
+   *
+   * @return The found macro, or null
+   */
+  private Macro getMacro(List<String> names) {
```

Comment:
This still looks very similar to `getFunctionsFrom()`. Is there a way to refactor common logic into a helper function?

Reply:
I'm not sure what else could be refactored, do you have any ideas?

Reply:
If macros can't have duplicate names like functions, then probably nothing of value.